src/worker/ingestion/action-matcher.ts

```diff
@@ -0,0 +1,490 @@
+            if (
+                !elements.some((element) => {
+                    if (step.href && element.href !== step.href) {
+                        return false // ELEMENT HREF IS A MISMATCH
+                    }
+                    if (step.tag_name && element.tag_name !== step.tag_name) {
+                        return false // ELEMENT TAG NAME IS A MISMATCH
+                    }
+                    if (step.text && element.text !== step.text) {
+                        return false // ELEMENT TEXT IS A MISMATCH
+                    }
+                    return true
+                })
+            ) {
+                // AT LEAST ONE ELEMENT MUST BE A SUBMATCH
+                return false
+            }
+        }
+        if (step.selector && !this.checkElementsAgainstSelector(elements, step.selector)) {
+            return false // SELECTOR IS A MISMATCH
+        }
+        return true
+    }
+
+    /**
+     * Sublevel 2 of action matching.
+     *
+     * Return whether the event is a match for the step's event name constraint.
+     * Step property: `event`.
+     */
+    private checkStepEvent(event: PluginEvent, step: ActionStep): boolean {
+        // CHECK CONDITIONS, OTHERWISE SKIPPED
+        if (step.event && event.event !== step.event) {
+            return false // EVENT NAME IS A MISMATCH
+        }
+        return true
+    }
+
+    /**
+     * Sublevel 2 of action matching.
+     *
+     * Return whether the event is a match for the step's fiter constraints.
+     * Step property: `properties`.
+     */
+    private async checkStepFilters(event: PluginEvent, elements: Element[], step: ActionStep): Promise<boolean> {
+        // CHECK CONDITIONS, OTHERWISE SKIPPED, OTHERWISE SKIPPED
+        if (step.properties && step.properties.length) {
+            // EVERY FILTER MUST BE A MATCH
+            for (const filter of step.properties) {
+                if (!(await this.checkEventAgainstFilter(event, elements, filter))) {
+                    return false
+                }
+            }
+        }
+        return true
+    }
+
+    /**
+     * Sublevel 3 of action matching.
+     */
+    private async checkEventAgainstFilter(
+        event: PluginEvent,
+        elements: Element[],
+        filter: PropertyFilter
+    ): Promise<boolean> {
+        let person: Person | undefined
+        switch (filter.type) {
+            case 'event':
+                return this.checkEventAgainstEventFilter(event, filter)
+            case 'person':
+                person = await this.db.fetchPerson(event.team_id, event.distinct_id)
+                return this.checkEventAgainstPersonFilter(person, filter)
+            case 'element':
+                return this.checkEventAgainstElementFilter(elements, filter)
+            case 'cohort':
+                person = await this.db.fetchPerson(event.team_id, event.distinct_id)
+                return await this.checkEventAgainstCohortFilter(person, filter)
```

Comment:
This `db.fetchPerson` could be fetching the same db rows multiple times per `ActionMatcher.match` in case we have a lot of steps that check user properties. 

It would be great if we could either recycle the person that we got while ingesting, or then just cache it here somehow.